avro-converter/src/main/java/io/confluent/copycat/avro/AvroData.java

```diff
@@ -0,0 +1,1073 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ **/
+
+package io.confluent.copycat.avro;
+
+import io.confluent.kafka.serializers.NonRecordContainer;
+
+import org.apache.avro.generic.GenericEnumSymbol;
+import org.apache.avro.generic.GenericFixed;
+import org.apache.avro.generic.GenericRecord;
+import org.apache.avro.generic.GenericRecordBuilder;
+import org.apache.avro.generic.IndexedRecord;
+import org.apache.kafka.copycat.data.Field;
+import org.apache.kafka.copycat.data.Schema;
+import org.apache.kafka.copycat.data.SchemaAndValue;
+import org.apache.kafka.copycat.data.SchemaBuilder;
+import org.apache.kafka.copycat.data.Struct;
+import org.apache.kafka.copycat.errors.DataException;
+import org.codehaus.jackson.JsonNode;
+import org.codehaus.jackson.node.ArrayNode;
+import org.codehaus.jackson.node.JsonNodeFactory;
+import org.codehaus.jackson.node.ObjectNode;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * Utilities for converting between our runtime data format and Avro, and (de)serializing that data.
+ */
+public class AvroData {
+  public static final String NAMESPACE = "io.confluent.copycat.avro";
+  // Avro does not permit empty schema names, which might be the ideal default since we also are
+  // not permitted to simply omit the name. Instead, make it very clear where the default is
+  // coming from.
+  public static final String DEFAULT_SCHEMA_NAME = "CopycatDefault";
+  public static final String DEFAULT_SCHEMA_FULL_NAME = NAMESPACE + "." + DEFAULT_SCHEMA_NAME;
+  public static final String MAP_ENTRY_TYPE_NAME = "MapEntry";
+  public static final String KEY_FIELD = "key";
+  public static final String VALUE_FIELD = "value";
+
+  public static final String COPYCAT_NAME_PROP = "copycat_name";
+  public static final String COPYCAT_DOC_PROP = "copycat_doc";
+  public static final String COPYCAT_VERSION_PROP = "copycat_version";
+  public static final String COPYCAT_DEFAULT_VALUE_PROP = "copycat_default";
+
+  public static final String COPYCAT_TYPE_PROP = "copycat_type";
+
+  public static final String COPYCAT_TYPE_INT8 = "int8";
+  public static final String COPYCAT_TYPE_INT16 = "int16";
+
+  public static final String AVRO_TYPE_UNION = NAMESPACE + ".Union";
+  public static final String AVRO_TYPE_ENUM = NAMESPACE + ".Enum";
+
+  public static final String AVRO_TYPE_ANYTHING = NAMESPACE + ".Anything";
+
+  private static final Map<String, Schema.Type> SPECIALIZED_TYPES_BY_TYPE_CODE = new HashMap<>();
```

Comment:
Ah! Those don't exist in Avro :)
NON_AVRO_TYPES is an ugly name, but probably clearer...